### ISSUE TITLE
[NERSC] numpy import error with desc-stack-weekly

### ISSUE THREAD

**Issue by humnaawan** [LSSTDESC/desc-help#56]
I cannot seem to be able to load numpy after loading `desc-stack-weekly`:
```
awan@cori19:~: python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly
Starting an environment for desc-stack-weekly now...
This may take a while. Be patient...
Including user python path: /global/homes/a/awan/.local/lib/python3.7/site-packages

(lsst-scipipe-0.4.1) awan@cori19:~: python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/__init__.py", line 140, in <module>
    from . import core
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.8 from "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-0.4.1/bin/python"
  * The NumPy version is: "1.19.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'
```
Things are fine if I load the desc-python (via `source /global/common/software/lsst/common/miniconda/setup_current_python.sh`).

Choose all applicable topics by placing an 'X' between the [ ]:
- [ ] jupyter
- [ ] Cori interactive command line
- [ ] Batch jobs
- [x] python
- [ ] CSCRATCH
- [ ] Community File System
- [ ] HPSS (tape)
- [ ] Data transfer and Globus
- [ ] New User Account or account access problems

**To Reproduce**
1. Run `python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly`
2. Open a python shell and try to import numpy.

**Comment by heather999** [comment]
Hi @humnaawan 
I think this part of the error output is the clue to what is going on:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/__init__.py", line 140, in <module>
    from . import core
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 48, in <module>
    raise ImportError(msg)
```
Your environment is picking up a local installation of numpy and likely other packages that are located under your $HOME directory.   I recall you needed to have that local install of CCL while we waited for the new release. Do you have an environment variable `$DESCPYTHONPATH` set which is pointing to your $HOME or possibly `$DESCSTACKUSERENV` ?  I would make sure those are unset and try again.

In my case, desc-stack-weekly's numpy seems fine:
```
heatherk@cori04:~> bash
heatherk@cori04:~> python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly
Starting an environment for desc-stack-weekly now...
This may take a while. Be patient...
(lsst-scipipe-0.4.1) bash-4.2$ python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> 
```

**Comment by humnaawan** [comment]
Hi @heather999 for the CCL issue, I didn't change any of my paths; I just ended up importing the local CCL repo in python. I do have a variable `$DESCPYTHONPATH` (`export DESCPYTHONPATH=/global/homes/a/awan/.local/lib/python3.7/site-packages`) in my `bash_profile.ext`. If I comment that out though, and try to load the desc-stack-weekly, I get:
```
Starting an environment for desc-stack-weekly now...
This may take a while. Be patient...
Traceback (most recent call last):
  File "/global/common/software/lsst/common/miniconda/start-kernel-cli.py", line 62, in <module>
    subprocess.call(command)
  File "/global/common/software/lsst/common/miniconda/py3.7-4.7.12.1-v2/envs/desc/lib/python3.7/subprocess.py", line 339, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/global/common/software/lsst/common/miniconda/py3.7-4.7.12.1-v2/envs/desc/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/global/common/software/lsst/common/miniconda/py3.7-4.7.12.1-v2/envs/desc/lib/python3.7/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'shifter': 'shifter'
```

**Comment by heather999** [comment]
Hi @humnaawan 
This new error is different and a little surprising, since shifter is a NERSC application that should be available on every Cori machine - unless your $PATH is somehow not set up as expected:
`FileNotFoundError: [Errno 2] No such file or directory: 'shifter': 'shifter'`
Can you start a fresh Cori login session and do:
`which shifter`  ?
In my case that looks like:
```
heatherk@cori04:~> which shifter
/usr/bin/shifter
```
If you see something similar in your new Cori window, I would go ahead and look at the output for:
`which python`
that would typically be `/usr/bin/python`
If that looks like what you have, I'd go ahead and try to start desc-stack-weekly again to see what happens:

`python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly`

**Comment by humnaawan** [comment]
hi @heather999 im sorry - the last error message was from a terminal in jupyterlab; i'm sorry i didnt realize there might be a difference between that vs. a direct terminal. if i run `which shifter` in a new jupyterlab terminal, i get what you have:
```
awan@cori19:/global/u2/a/awan: which shifter
/usr/bin/shifter
```
for python, however, i get something different:
```
awan@cori19:/global/u2/a/awan: which python
/usr/common/software/python/3.7-anaconda-2019.07/bin/python
```
however, i'm no longer getting the shifter error when o run `python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly` so it might have had something to do with sourcing the updated bash profile and then trying to load the stack in the same terminal as before.

now i get the numpy error still, both in a jupyterlab terminal and a cori termins. here's the output from the cori terminal:
```
awan@cori09:~: python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly
Starting an environment for desc-stack-weekly now...
This may take a while. Be patient...
Including user python path: /global/homes/a/awan/.local/lib/python3.7/site-packages
(lsst-scipipe-0.4.1) awan@cori09:~: python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/__init__.py", line 140, in <module>
    from . import core
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.8 from "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-0.4.1/bin/python"
  * The NumPy version is: "1.19.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'
```

here are the outputs for the shifter and python checks in the cori terminal (before loading the stack):
```
awan@cori09:~: which python
/usr/common/software/python/3.7-anaconda-2019.07/bin/python
awan@cori09:~: which shifter
/usr/bin/shifter
```

**Comment by johannct** [comment]
.local is really terrible...... Humna do you intend to use 3.8 or 3.7? stack weekly clearly sets you up for 3.8. By putting a (DESC)PYTHONPATH to bash_profile.ext you kinda commit yourself to use version 3.7 forever after..... which is incompatible now with your attempt to you use a recent stack setup....

**Comment by humnaawan** [comment]
i dont really mind either 3.7 or 3.8. here's what i have in my bash_profile re the python path:
```
export PYTHONPATH=/global/homes/a/awan/.local/bin:$PYTHONPATH
```
do you suggest removing that?

**Comment by johannct** [comment]
above you said you had a DESCPYTHONPATH set in your bash_profile.ext.....

**Comment by humnaawan** [comment]
yes, sorry - that was a typo. the python path is also in bash_profile.ext.

**Comment by johannct** [comment]
so you mean that you have /global/homes/a/awan/.local/bin added to PYTHONPATH and /global/homes/a/awan/.local/lib/python3.7/site-packages added to DESCPYTHONPATH?

**Comment by humnaawan** [comment]
yes, that was the case before. im getting the numpy error issue even if i comment out both the `DESCPYTHONPATH` and `PYTHONPATH` lines in `bash_profile.ext`.

specifically, with neither `DESCPYTHONPATH` and `PYTHONPATH` set in the `bash_profile.ext`, in a new cori terminal, i have
```
awan@cori09:~: which python
/usr/common/software/python/3.7-anaconda-2019.07/bin/python
awan@cori09:~: python
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
```
and things work. however, the error appears when i do:
```
awan@cori09:~: python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly
Starting an environment for desc-stack-weekly now...
This may take a while. Be patient...
Including user python path: /global/homes/a/awan/.local/lib/python3.7/site-packages
(lsst-scipipe-0.4.1) awan@cori09:~: which python
/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-0.4.1/bin/python
(lsst-scipipe-0.4.1) awan@cori09:~: python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/__init__.py", line 140, in <module>
    from . import core
  File "/global/homes/a/awan/.local/lib/python3.7/site-packages/numpy/core/__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.8 from "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-0.4.1/bin/python"
  * The NumPy version is: "1.19.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'
```

**Comment by johannct** [comment]
ok so how come that your python is pointing to an anaconda version? On first login to a cori machine the python in my path is johannct@cori11:~> which python
/usr/bin/python

so you must be doing something to have /usr/common/software/python/3.7-anaconda-2019.07/bin/python first in your path?

**Comment by johannct** [comment]
are you sourcing or launching /usr/common/software/python/3.7-anaconda-2019.07/bin/activate somewhere?

**Comment by heather999** [comment]
Let's take a breath :) 
We do need to document this case when people start up a terminal from Jupyter - because the situation is a little different.  I can confirm though that starting up a jupyter terminal and setting up desc-stack-weekly as you have been trying to do, should work.  I pasted the output when I do that at the bottom of this post. The situation with the active version of python is different, so while if you were logged into a typical Cori node your default python would be `/usr/bin/python`, when you start up a terminal, your default python is whatever is set up for jupyterhub which is this ` /usr/common/software/python/3.7-anaconda-2019.07/bin/python`  But that's fine.

```
heatherk@cori14:/global/u1/h/heatherk> python --version
Python 3.8.5
heatherk@cori14:/global/u1/h/heatherk> which conda
/global/common/cori_cle7/software/jupyter/cori/20-12/bin/conda
heatherk@cori14:/global/u1/h/heatherk> conda list envs
# packages in environment at /global/common/cori_cle7/software/jupyter/cori/20-12:
#
# Name                    Version                   Build  Channel
heatherk@cori14:/global/u1/h/heatherk> which shifter
/usr/bin/shifter
heatherk@cori14:/global/u1/h/heatherk> echo $SHELL
/bin/bash
heatherk@cori14:/global/u1/h/heatherk> bash
heatherk@cori14:/global/u1/h/heatherk> python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly
Starting an environment for desc-stack-weekly now...
This may take a while. Be patient...
(lsst-scipipe-0.4.1) bash-4.2$ which python
/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-0.4.1/bin/python
(lsst-scipipe-0.4.1) bash-4.2$ python --version
Python 3.8.6
(lsst-scipipe-0.4.1) bash-4.2$ python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> ```

Now we need to figure out what's going on with Humna's env.. let me take a moment to write a coherent response.

**Comment by humnaawan** [comment]
i mean this is the content of my `bash_profile.ext`:
```
awan@cori09: less ~/.bash_profile.ext 

# then
#   export MY_ENV_VAR="my_val"
# fi

# module load python/2.7-anaconda
#module load python/3.7-anaconda-2019.07

alias gs="git status"
alias gla="git log"
alias gl="git log --oneline"
alias emacs="emacs -nw"

alias ls="ls --color=auto"
export PS1='\[\033[01;36m\]\u@\h\[\033[33m\]:\[\033[01;33m\]\w:\[\033[00m\] '
export CLICOLOR=1
#export LS_COLORS=ExFxCxDxBxegedabagacad
export LS_COLORS="ow=01;90:di=01;31"

# Jan 31, 2018:
# --
#alias loadlsst="source ~/lsst/loadLSST.bash ; setup sims_maf_contrib -t $USER -t sims"  # updated Jan 17
#alias loadlsst="source /global/common/software/lsst/cori-haswell-gcc/stack/setup_w_2017_46_py3_gcc6.sh ; setup sims_maf -t $USER"
# Nov 8, 2020
#alias loadlsst="source /global/common/software/lsst/cori-haswell-gcc/stack/setup_current_sims.sh ; setup sims_maf_contrib -r /global/homes/a/awan/LSST/lsstRepos/sims_maf_contrib ; setup sims_maf -r /global/homes/a/awan/LSST/lsstRepos/sims_maf"
alias loadlsst="python /global/common/software/lsst/common/miniconda/start-kernel-cli.py desc-stack-weekly ; setup sims_maf_contrib -r /global/homes/a/awan/LSST/lsstRepos/sims_maf_contrib ; setup sims_maf -r /global/homes/a/awan/LSST/lsstRepos/sims_maf"

# June 28, 2018
#export PYTHONPATH=/global/homes/a/awan/SO/software/CAMB-0.1.6.1:$PYTHONPATH
#export PYTHONPATH=/global/homes/a/awan/SO/software/CAMB-0.1.6.1/pycamb:$PYTHONPATH
#export PYTHONPATH=/global/homes/a/awan/SO/:$PYTHONPATH

#export PYTHONPATH=/global/homes/a/awan/.local/bin:$PYTHONPATH

# Jan 24, 2020
#export DESCPYTHONPATH=/global/homes/a/awan/.local/lib/python3.7/site-packages
# oct 14, 2020
#export DESCPYTHONPATH=/global/homes/a/awan/.local/cori/3.7-anaconda-2019.07/lib/python3.7/site-packages/

# june 3, 2020
#export PYTHONPATH=/global/homes/a/awan/so/:$PYTHONPATH

# oct 20, 2020
alias loaddesc="source /global/common/software/lsst/common/miniconda/setup_current_python.sh"

[ -e $HOME/.dbgdot ] && echo "exiting .bash_profile.ext" >&2

# end .bash_profile.ext
```
am i missing something?

**Comment by heather999** [comment]
That specific message:
`Including user python path: /global/homes/a/awan/.local/lib/python3.7/site-packages`
would be displayed when $DESCPYTHONPATH  is set.  
The contents of both `.bashrc.ext` and `.bashrc` matter. At the command line, can you do:

`set | grep DESC` and see what output there is?  There is probably still some setting of DESCPYTHONPATH - possibly in your $HOME/.bashrc file.

**Comment by humnaawan** [comment]
okay i have:
```
awan@cori09:~: set | grep DESC
DESCPYTHONPATH=/global/homes/a/awan/.local/lib/python3.7/site-packages
```
looking at `.bashrc.ext`, i found the culprit: i was setting up the DESCPYTHONPATH there too for some reason. i no longer have a problem with importing numpy with the desc-stack-weekly.

i'm really sorry for all this - i set up these things a while ago when i was much less careful re what i was doing. thanks so much for all your help!

**Comment by johannct** [comment]
So lesson learned is really : no python config in the .bash_<whatever> files and if at all possible avoid .local like the plague

**Comment by heather999** [comment]
No worries @humnaawan - I think this happens a lot.  We set up our .bashrc and come back weeks/months later and there is a need to adjust things.
Well - I think the lesson is that people will set things up in the .bash file for convenience and the informational message in start_kernel_cli needs to be updated to make it very clear that is associated with setting DESCPYTHONPATH.
Esp since desc-stack-weekly is a shifter image, your external environment including whatever python is active, is actually irrelevant.  We created this DESCPYTHONPATH to provide one way for users to install things locally, but while it is helpful in the short term - it often creates confusion later.

On our end - we have to work out a good line of questioning to help figure out what's going on.
